elastic/apm-server
# intake: Support non-blocking event processing

## Motivation/summary

Introduces a new bool `async` query parameter for `intake/v2/events`.
This allows clients to ask the APM Server to process the contents of the
request in an async manner. This will only happen if the APM Server has
enough capacity to service the request, otherwise it will return the
infamous `publish.ErrFull` error.

## Checklist

~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~

## How to test these changes

1. `docker-compose up -d`
2. `cd systemtest && go run cmd/runapm/main.go`
3. `export RUNAPM_PORT=REPLACE_WITH_RUNAPM_PORT`
3. `curl -H 'Content-type: application/x-ndjson' --data-binary @testdata/intake-v2/errors_transaction_id.ndjson "http://localhost:${RUNAPM_PORT}/intake/v2/events?async=true"`
4. `curl -H 'Content-type: application/x-ndjson' --data-binary @testdata/intake-v2/invalid-event-type.ndjson "http://localhost:${RUNAPM_PORT}/intake/v2/events?async=true"`, see error:
```
{
  "accepted": 0,
  "errors": [
    {
      "message": "tennis-court: did not recognize object type",
      "document": "{\"tennis-court\": {\"name\": \"Centre Court, Wimbledon\"}}"
    }
  ]
}
```
5. Check that `logs-apm.error-default/_search` contains 1 document.
6. `for i in {1..5}; do curl -H 'Content-type: application/x-ndjson' --data-binary @testdata/intake-v2/heavy.ndjson "http://localhost:${RUNAPM_PORT}/intake/v2/events?async=true"; done`. Observe at least 2 "queue is full" responses.
```
{
  "accepted": 0,
  "errors": [
    {
      "message": "queue is full"
    }
  ]
}
```

## Related issues

Closes #7872
